File: src/main/java/com/now/nowbot/service/OsuApiService/impl/DiscussionApiImpl.java
package com.now.nowbot.service.OsuApiService.impl;

import com.now.nowbot.model.JsonData.Discussion;
import com.now.nowbot.model.JsonData.DiscussionDetails;
import com.now.nowbot.service.OsuApiService.OsuDiscussionApiService;
import org.springframework.lang.Nullable;
import org.springframework.stereotype.Service;

import java.util.Objects;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicInteger;

@Service
public class DiscussionApiImpl implements OsuDiscussionApiService {
    OsuApiBaseService base;

    public DiscussionApiImpl(OsuApiBaseService baseService) {
        base = baseService;
    }

    //sort默认id_desc，即最新的在前。也可以是 id_asc
    @Override
    public Discussion getBeatMapDiscussion(
            Long bid,
            Long sid,
            @Nullable
            BeatMapSetStatus status,
            int limit,
            @Nullable
            DiscussionDetails.MessageType[] types,
            @Nullable
            Boolean onlyResolved,
            int page,
            @Nullable
            String sort,
            @Nullable
            Long uid
    ) {
        AtomicInteger pageAtm = new AtomicInteger(page);
        Discussion discussion =
                getBeatMapDiscussion(bid, sid, status, limit, types, onlyResolved, page, sort, uid, null);
        while (Objects.nonNull(discussion.getCursorString()) && pageAtm.get() < 10) {
            var other =
                    getBeatMapDiscussion(bid, sid, status, limit, types, onlyResolved, pageAtm.getAndAdd(1), sort, uid, null);
            discussion.setCursorString(other.getCursorString());
            discussion.mergeDiscussion(other, sort);
        }
        return discussion;
    }

    public Discussion getBeatMapDiscussion(
            Long bid,
            Long sid,
            @Nullable
            BeatMapSetStatus status,
            int limit,
            @Nullable
            DiscussionDetails.MessageType[] types,
            @Nullable
            Boolean onlyResolved,
            int page,
            @Nullable
            String sort,
            Long uid,
            String cursor
    ) {
        return base.osuApiWebClient.get()
                .uri(u -> u.path("beatmapsets/discussions")
                        .queryParamIfPresent("beatmap_id", Optional.ofNullable(bid))
                        .queryParamIfPresent("beatmapset_id", Optional.ofNullable(sid))
                        .queryParamIfPresent("beatmapset_status", Optional.ofNullable(status))
                        .queryParam("limit", limit)
                        .queryParamIfPresent("message_types[]", Optional.ofNullable(types))
                        .queryParamIfPresent("only_resolved", Optional.ofNullable(onlyResolved))
                        .queryParam("page", page)
                        .queryParamIfPresent("sort", Optional.ofNullable(sort))
                        .queryParamIfPresent("user", Optional.ofNullable(uid))
                        .queryParamIfPresent("cursor_string", Optional.ofNullable(cursor))
                        .build())
                .headers(base::insertHeader)
                .retrieve()
                .bodyToMono(Discussion.class)
                .block();
    }

    public enum BeatMapSetStatus {
        all, ranked, qualified, disqualified, never_qualified,
    }
}

File: src/main/java/com/now/nowbot/service/osuApiService/impl/DiscussionApiImpl.kt
package com.now.nowbot.service.osuApiService.impl

import com.now.nowbot.model.osu.Discussion
import com.now.nowbot.service.osuApiService.OsuDiscussionApiService
import org.springframework.http.HttpHeaders
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.util.UriBuilder
import java.util.*

@Service
class DiscussionApiImpl(var base: OsuApiBaseService) : OsuDiscussionApiService {
    //sort默认id_desc，即最新的在前。也可以是 id_asc
    override fun getBeatMapDiscussion(
        bid: Long?,
        sid: Long?,
        status: BeatMapSetStatus?,
        limit: Int?,
        types: List<String>?,
        onlyResolved: Boolean?,
        page: Int?,
        sort: String?,
        uid: Long?
    ): Discussion {
        var count = 0
        val discussion =
            getBeatMapDiscussion(bid, sid, status, limit, types, onlyResolved, page, sort, uid, null)
        while (discussion.cursorString.isNullOrEmpty().not() && count++ < 10) {
            val other = getBeatMapDiscussion(
                bid, sid, status, limit, types, onlyResolved, page, sort, uid, discussion.cursorString
            )

            discussion.cursorString = other.cursorString
            discussion.mergeDiscussion(other, sort!!)
        }
        return discussion
    }

    fun getBeatMapDiscussion(
        bid: Long?,
        sid: Long?,
        status: BeatMapSetStatus?,
        limit: Int?,
        types: List<String>?,
        onlyResolved: Boolean?,
        page: Int?,
        sort: String?,
        uid: Long?,
        cursor: String?
    ): Discussion {
        return base.request { client: WebClient ->
            client
                .get()
                .uri { u: UriBuilder ->
                    u.path("beatmapsets/discussions")
                        .queryParamIfPresent("beatmap_id", Optional.ofNullable(bid))
                        .queryParamIfPresent("beatmapset_id", Optional.ofNullable(sid))
                        .queryParamIfPresent("beatmapset_status", Optional.ofNullable(status))
                        .queryParamIfPresent("limit", Optional.ofNullable(limit))
                        .queryParamIfPresent("message_types[]", Optional.ofNullable(types?.toTypedArray()))
                        .queryParamIfPresent("only_resolved", Optional.ofNullable(onlyResolved))
                        .queryParamIfPresent("page", Optional.ofNullable(page))
                        .queryParamIfPresent("sort", Optional.ofNullable(sort))
                        .queryParamIfPresent("user", Optional.ofNullable(uid))
                        .queryParamIfPresent("cursor_string", Optional.ofNullable(cursor))
                        .build()
                }
                .headers { headers: HttpHeaders? ->
                    base.insertHeader(
                        headers!!
                    )
                }
                .retrieve()
                .bodyToMono(Discussion::class.java)
        }
    }

    enum class BeatMapSetStatus {
        all, ranked, qualified, disqualified, never_qualified,
    }
}

